### PR TITLE
extracting the substructure patterns of CircularFingerprint. 

### PR DESCRIPTION
**Pull request by vedina**

### CODE REVIEW

File: descriptor/fingerprint/src/main/java/org/openscience/cdk/fingerprint/CircularFingerprinter.java [hunk 1]
```diff
@@ -1258,6 +1262,14 @@ private int findBond(int a1, int a2) {
             if (atomAdj[a1][n] == a2) return bondAdj[a1][n];
         return -1;
     }
+    
+    // convecience: find value within array
+    private int findArrayIndex(int value, int array[]){
```
Comment (johnmay) [inline descriptor/fingerprint/src/main/java/org/openscience/cdk/fingerprint/CircularFingerprinter.java]: Can use ```Ints.indexOf()```

**Comment by johnmay** [comment]
I need to rebase this but looks like interesting functionality. Late here but the only things that stick out a little at the moment is the SMARTS generation in the fingerprint, okay for now but I would prefer to centralise this but obviously would need to write that/use AMBITs one. Late here but wondering how you get rings in a circular fingerprint? Shouldn't all the subgraphs be trees, will probably become clear once I look more in depth.

**Comment by vedina** [comment]
* Fine for centralize SMARTS generation if possible, as here it uses internal CircularFingerprint structures;
* There are no AMBIT dependencies in this code (intentionally) - of course SMARTS should be valid at the end;

* Rings appear if the depth is large enough. @ntk73 may explain implementation details.
As this feature is modeled after [JCompoundMapper](http://jcompoundmapper.sourceforge.net/) here is JCompoundMapper output for benzene

ECFP6  ( :1 means count , it's not part of the SMARTS)
````
[*]=1=C=1C=C2[*]2:1	
[*]=1=C=12([*]2):1	
[*]1C1=2(=[*]=2):1	
[*]1C1=CC=CC=2=[*]=2:1	
[*]=1=C=1C=CC=C2[*]2:1	
c1ccccc1:1	
[*]1c1=CC=Cc=2=[*]=2:1	
[*]1c1=CC=2=[*]=2:1	
[*]1C1=CC=CC=2=[*]=2:1
````

ECFP4  (no ring structures)
````
[*]1C1=2(=[*]=2):1	
[*]=1=C=1C=CC=C2[*]2:1	
[*]1C1=CC=CC=2=[*]=2:1	
[*]1C1=CC=2=[*]=2:1	
[*]=1=C=1C=C2[*]2:1	
[*]=1=C=12([*]2):1
````

File: descriptor/fingerprint/src/main/java/org/openscience/cdk/fingerprint/CircularFingerprinter.java (hunk 1 above)
Comment (vedina) [inline descriptor/fingerprint/src/main/java/org/openscience/cdk/fingerprint/CircularFingerprinter.java]: agree

**Comment by johnmay** [comment]
Rebased - #231

**Pull request closed without merging**